nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceResolver.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.nifi.authorization.user.NiFiUser;
+import org.apache.nifi.controller.flow.FlowManager;
+import org.apache.nifi.flow.ControllerServiceAPI;
+import org.apache.nifi.flow.ExternalControllerServiceReference;
+import org.apache.nifi.flow.VersionedConfigurableExtension;
+import org.apache.nifi.flow.VersionedControllerService;
+import org.apache.nifi.flow.VersionedProcessGroup;
+import org.apache.nifi.flow.VersionedProcessor;
+import org.apache.nifi.flow.VersionedPropertyDescriptor;
+import org.apache.nifi.groups.ProcessGroup;
+import org.apache.nifi.registry.flow.FlowSnapshotContainer;
+import org.apache.nifi.registry.flow.RegisteredFlowSnapshot;
+import org.apache.nifi.registry.flow.mapping.NiFiRegistryFlowMapper;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+public class StandardControllerServiceResolver implements ControllerServiceResolver {
+
+    private final Authorizer authorizer;
+    private final FlowManager flowManager;
+    private final NiFiRegistryFlowMapper flowMapper;
+    private final ControllerServiceProvider controllerServiceProvider;
+    private final ControllerServiceApiLookup controllerServiceApiLookup;
+
+    public StandardControllerServiceResolver(final Authorizer authorizer,
+                                             final FlowManager flowManager,
+                                             final NiFiRegistryFlowMapper flowMapper,
+                                             final ControllerServiceProvider controllerServiceProvider,
+                                             final ControllerServiceApiLookup controllerServiceApiLookup) {
+        this.authorizer = authorizer;
+        this.flowManager = flowManager;
+        this.flowMapper = flowMapper;
+        this.controllerServiceProvider = controllerServiceProvider;
+        this.controllerServiceApiLookup = controllerServiceApiLookup;
+    }
+
+    @Override
+    public void resolveInheritedControllerServices(final FlowSnapshotContainer flowSnapshotContainer, final String parentGroupId, final NiFiUser user) {
+        final RegisteredFlowSnapshot topLevelSnapshot = flowSnapshotContainer.getFlowSnapshot();
+        final VersionedProcessGroup versionedGroup = topLevelSnapshot.getFlowContents();
+        final Map<String, ExternalControllerServiceReference> externalControllerServiceReferences = topLevelSnapshot.getExternalControllerServices();
+
+        final ProcessGroup parentGroup = flowManager.getGroup(parentGroupId);
+
+        final Set<VersionedControllerService> ancestorServices = parentGroup.getControllerServices(true).stream()
+                .filter(serviceNode -> serviceNode.isAuthorized(authorizer, RequestAction.READ, user))
+                .map(serviceNode -> flowMapper.mapControllerService(serviceNode, controllerServiceProvider, Collections.emptySet(), Collections.emptyMap()))
```

Comment:
When importing a versioned group with a child versioned group, I get the following exception:

```
2023-05-04 07:03:09,458 ERROR [NiFi Web Server-2125] o.a.nifi.web.api.config.ThrowableMapper An unexpected error has occurred: java.lang.UnsupportedOperationException. Returning Internal Server Error response.
java.lang.UnsupportedOperationException: null
        at java.base/java.util.AbstractMap.put(AbstractMap.java:209)
        at org.apache.nifi.registry.flow.mapping.NiFiRegistryFlowMapper.mapPropertyDescriptors(NiFiRegistryFlowMapper.java:638)
        at org.apache.nifi.registry.flow.mapping.NiFiRegistryFlowMapper.mapControllerService(NiFiRegistryFlowMapper.java:532)
        at org.apache.nifi.controller.service.StandardControllerServiceResolver.lambda$resolveInheritedControllerServices$1(StandardControllerServiceResolver.java:72)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.apache.nifi.controller.service.StandardControllerServiceResolver.resolveInheritedControllerServices(StandardControllerServiceResolver.java:73)
        at org.apache.nifi.web.StandardNiFiServiceFacade.resolveInheritedControllerServices(StandardNiFiServiceFacade.java:4062)
```